File: C++/Leetcode 1071.cpp
#include <stdio.h>
#include <bits/stdc++.h>
using namespace std;

class Solution {
    bool Helper(string Parent, string Child)
    {
        string temp_child = Child;
        while (Child.length() != Parent.length())
        {
            Child += temp_child;
        }
        if (Child.compare(Parent) == 0)
        {
            return true;
        }
        return false;
    }
public:
    string gcdOfStrings(string str1, string str2) {
        if (str1.length() < str2.length())
        {
            string temp = str1;
            str1 = str2;
            str2 = temp;
        }

        //  Str1 is longer string now

        for (int i = str2.length()-1; i >= 0; i--)
        {
            int substring_length = i+1;

            if (str2.length() % substring_length == 0 && str1.length() % substring_length == 0)
            {
                string sub = str2.substr(0,substring_length);

                if (Helper(str2,sub) && Helper(str1,sub))
                {
                    return sub;
                }
            }
        }
        return "";
    }
};

File: Python/Leetcode 1071.py
class Solution:
    def Helper(self, str1: str, str2: str) -> bool:
        temp = str2

        while (len(str2) != len(str1)):
            str2 = str2 + temp
        
        return str2 == str1


    def gcdOfStrings(self, str1: str, str2: str) -> str:
        length_str1 = len(str1)
        length_str2 = len(str2)

        if (length_str1 < length_str2):
            temp = str2
            str1 = str2
            str2 = temp

        for index in range(len(str2)-1,-1,-1):
            substring_length = index+1

            if length_str1 % substring_length == 0 and length_str2 % substring_length == 0:
                print(substring_length)
                sub = str2[0:substring_length]
        
                if (self.Helper(str1,sub) and self.Helper(str2,sub)):
                    return sub
        
        return ""
